merkle/log_verifier_test.go

```diff
@@ -0,0 +1,448 @@
+				{dh("ca854ea128ed050b41b35ffc1b87b8eb2bde461e9e3b5596ece6b9d5975a0ae0"), 32},
+				{dh("d37ee418976dd95753c1c73862b9398fa2a2cf9b4ff0fdfe8b30cd95209614b7"), 32}}},
+		{2,
+			5,
+			2,
+			[3]logProofTestVector{
+				{dh("5f083f0a1a33ca076a95279832580db3e0ef4584bdff1f54c8a360f50de3031e"), 32},
+				{dh("bc1a0643b12e4d2d7c77918f44e0f4f79a838b6cf9ec5b5c283e1f4d88599e6b"), 32},
+				{dh(""), 0}}},
+	}
+}
+
+func TestVerifyInclusionProof(t *testing.T) {
+	v := getVerifier()
+	path := [][]byte{}
+	// Various invalid paths
+	if err := v.VerifyInclusionProof(0, 0, path, []byte{}, []byte{}); err == nil {
+		t.Fatal("Incorrectly verified invalid path 1")
+	}
+	if err := v.VerifyInclusionProof(0, 1, path, []byte{}, []byte{}); err == nil {
+		t.Fatal("Incorrectly verified invalid path 2")
+	}
+	if err := v.VerifyInclusionProof(1, 0, path, []byte{}, []byte{}); err == nil {
+		t.Fatal("Incorrectly verified invalid path 3")
+	}
+	if err := v.VerifyInclusionProof(2, 1, path, []byte{}, []byte{}); err == nil {
+		t.Fatal("Incorrectly verified invalid path 4")
+	}
+
+	if err := v.VerifyInclusionProof(0, 0, path, dh(sha256EmptyTreeHash), []byte{}); err == nil {
+		t.Fatal("Incorrectly verified invalid root 1")
+	}
+	if err := v.VerifyInclusionProof(0, 1, path, dh(sha256EmptyTreeHash), []byte{}); err == nil {
+		t.Fatal("Incorrectly verified invalid root 2")
+	}
+	if err := v.VerifyInclusionProof(1, 0, path, dh(sha256EmptyTreeHash), []byte{}); err == nil {
+		t.Fatal("Incorrectly verified invalid root 3")
+	}
+	if err := v.VerifyInclusionProof(2, 1, path, dh(sha256EmptyTreeHash), []byte{}); err == nil {
+		t.Fatal("Incorrectly verified invalid root 4")
+	}
+
+	// Known good paths.
+
+	inclusionProofs := getInclusionTestVector()
+	inputs := getInputs()
+	roots := getRoots()
+	// i = 0 is an invalid path.
+	for i := 1; i < 6; i++ {
+		// Construct the path.
+		proof := [][]byte{}
+		for j := int64(0); j < inclusionProofs[i].proofLength; j++ {
+			proof = append(proof, inclusionProofs[i].proof[j].h)
+		}
+		err := verifierCheck(&v, inclusionProofs[i].leaf-1, inclusionProofs[i].snapshot, proof,
+			roots[inclusionProofs[i].snapshot-1].h,
+			inputs[inclusionProofs[i].leaf-1].h)
+		if err != nil {
+			t.Fatalf("i=%d: %s", i, err)
+		}
+	}
+
+}
+
+func TestVerifyConsistencyProof(t *testing.T) {
+	v := getVerifier()
+
+	proof := [][]byte{}
+	root1 := []byte("don't care")
+	root2 := []byte("don't care")
+
+	// Snapshots that are always consistent
+	if err := verifierConsistencyCheck(&v, 0, 0, root1, root2, proof); err != nil {
+		t.Fatalf("Failed to verify proof: %s", err)
+	}
+	if err := verifierConsistencyCheck(&v, 0, 1, root1, root2, proof); err != nil {
+		t.Fatalf("Failed to verify proof: %s", err)
+	}
+	if err := verifierConsistencyCheck(&v, 1, 1, root1, root2, proof); err != nil {
+		t.Fatalf("Failed to verify proof: %s", err)
+	}
+
+	// Invalid consistency proofs.
+	// Time travel to the past.
```

Comment:
Can I see the design doc for the time machine? And has there been a safety audit?